ghc-toolkit/boot-libs/asterius-prelude/src/Asterius/UTF8.hs

```diff
@@ -0,0 +1,48 @@
+module Asterius.UTF8
+  ( utf8ToJSString,
+    utf8FromJSString,
+  )
+where
+
+import Asterius.Magic
+import Asterius.Prim
+import qualified Data.ByteString.Internal as BS
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.ByteString.Unsafe as BS
+import Foreign
+
+{-# INLINE utf8ToJSString #-}
+utf8ToJSString :: LBS.ByteString -> JSString
+utf8ToJSString s
+  | LBS.null s = js_str_empty
+  | otherwise = accursedUnutterablePerformIO $ do
+    dec <- js_dec
+    LBS.foldrChunks
+      (\c m -> BS.unsafeUseAsCStringLen c (uncurry (js_dec_chunk dec)) *> m)
+      (pure ())
+      s
+    js_dec_result dec
+
+{-# INLINE utf8FromJSString #-}
+utf8FromJSString :: JSString -> LBS.ByteString
+utf8FromJSString s = accursedUnutterablePerformIO $ do
+  l <- js_str_len s
+  if l == 0
+    then pure mempty
+    else fmap LBS.fromStrict $ BS.createUptoN (l * 3) $ \p ->
+      js_utf8_from_str s p (l * 3)
+
+foreign import javascript "''" js_str_empty :: JSString
```

Comment:
Do we really have to `foreign import` this? Don't we have a way to create a JSString more directly?

Reply:
To get a `JSVal` in the Haskell heap we need to call a `foreign import javascript` anyway, I don't think there's a more direct approach.